// js-to-ts-converter/dist/converter/add-class-property-declarations/add-class-property-declarations.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const ts_morph_1 = require("ts-morph");
const parse_js_classes_1 = require("./parse-js-classes");
const correct_js_properties_1 = require("./correct-js-properties");
const logger_1 = require("../../logger/logger");
/**
 * Parses all source files looking for ES6 classes, and takes any `this`
 * property access to create a PropertyDeclaration for the class.
 *
 * For example:
 *
 *     class Something {
 *         constructor() {
 *             this.someProp = 1;
 *         }
 *     }
 *
 * Is changed to:
 *
 *     class Something {
 *         someProp: any;
 *
 *         constructor() {
 *             this.someProp = 1;
 *         }
 *     }
 */
function addClassPropertyDeclarations(tsAstProject) {
    // Parse the JS classes for all of the this.xyz properties that they use
    const jsClasses = parse_js_classes_1.parseJsClasses(tsAstProject);
    // Correct the JS classes' properties for superclass/subclass relationships
    // (essentially remove properties from subclasses that are defined by their
    // superclasses)
    const propertiesCorrectedJsClasses = correct_js_properties_1.correctJsProperties(jsClasses);
    // Fill in field definitions for each of the classes
    propertiesCorrectedJsClasses.forEach(jsClass => {
        const sourceFile = tsAstProject.getSourceFileOrThrow(jsClass.path);
        logger_1.default.verbose(`  Updating class '${jsClass.name}' in '${sourceFile.getFilePath()}'`);
        const classDeclaration = sourceFile.getClassOrThrow(jsClass.name);
        const jsClassProperties = jsClass.properties;
        // If the utility was run against a TypeScript codebase, we should not
        // fill in property declarations for properties that are already
        // declared in the class. However, we *should* fill in any missing
        // declarations. Removing any already-declared declarations from the
        // jsClassProperties.
        const currentPropertyDeclarations = classDeclaration.getInstanceProperties()
            .reduce((props, prop) => {
            const propName = prop.getName();
            return propName ? props.add(propName) : props;
        }, new Set());
        const undeclaredProperties = [...jsClassProperties]
            .filter((propName) => !currentPropertyDeclarations.has(propName));
        // Add all currently-undeclared properties
        const propertyDeclarations = undeclaredProperties.map(propertyName => {
            return {
                name: propertyName,
                type: 'any',
                scope: ts_morph_1.Scope.Public
            };
        });
        logger_1.default.verbose(`    Adding property declarations for properties: '${undeclaredProperties.join("', '")}'`);
        classDeclaration.insertProperties(0, propertyDeclarations);
    });
    return tsAstProject;
}
exports.addClassPropertyDeclarations = addClassPropertyDeclarations;
